src/libraries/System.Private.CoreLib/src/System/Threading/Semaphore.Windows.cs

```diff
@@ -12,43 +11,120 @@ public sealed partial class Semaphore
+        private unsafe void CreateSemaphoreCore(
+            int initialCount,
+            int maximumCount,
+            string? name,
+            NamedWaitHandleOptionsInternal options,
+            out bool createdNew)
         {
-            Debug.Assert(initialCount >= 0);
-            Debug.Assert(maximumCount >= 1);
-            Debug.Assert(initialCount <= maximumCount);
+            ArgumentOutOfRangeException.ThrowIfNegative(initialCount);
+            ArgumentOutOfRangeException.ThrowIfNegativeOrZero(maximumCount);
 
-#if TARGET_UNIX || TARGET_BROWSER || TARGET_WASI
+            if (initialCount > maximumCount)
+            {
+                throw new ArgumentException(SR.Argument_SemaphoreInitialMaximum);
+            }
+
+#if !TARGET_WINDOWS
             if (name != null)
+            {
                 throw new PlatformNotSupportedException(SR.PlatformNotSupported_NamedSynchronizationPrimitives);
+            }
 #endif
-            SafeWaitHandle myHandle = Interop.Kernel32.CreateSemaphoreEx(IntPtr.Zero, initialCount, maximumCount, name, 0, AccessRights);
 
-            int errorCode = Marshal.GetLastPInvokeError();
-            if (myHandle.IsInvalid)
+            void* securityAttributesPtr = null;
+            SafeWaitHandle myHandle;
+            int errorCode;
+#if TARGET_WINDOWS
+            Thread.CurrentUserSecurityDescriptorInfo securityDescriptorInfo = default;
+            Interop.Kernel32.SECURITY_ATTRIBUTES securityAttributes = default;
+            if (!string.IsNullOrEmpty(name) && options.WasSpecified)
             {
-                myHandle.Dispose();
+                name = options.GetNameWithSessionPrefix(name);
+                if (options.CurrentUserOnly)
+                {
+                    securityDescriptorInfo = new(CurrentUserOnlyAceRights);
+                    securityAttributes.nLength = (uint)sizeof(Interop.Kernel32.SECURITY_ATTRIBUTES);
+                    securityAttributes.lpSecurityDescriptor = (void*)securityDescriptorInfo.SecurityDescriptor;
+                    securityAttributesPtr = &securityAttributes;
+                }
+            }
 
-                if (!string.IsNullOrEmpty(name) && errorCode == Interop.Errors.ERROR_INVALID_HANDLE)
-                    throw new WaitHandleCannotBeOpenedException(
-                        SR.Format(SR.Threading_WaitHandleCannotBeOpenedException_InvalidHandle, name));
+            using (securityDescriptorInfo)
+            {
+#endif
+                myHandle =
+                    Interop.Kernel32.CreateSemaphoreEx(
+                        (nint)securityAttributesPtr,
+                        initialCount,
+                        maximumCount,
+                        name,
+                        flags: 0,
+                        AccessRights);
+                errorCode = Marshal.GetLastPInvokeError();
+
+                if (myHandle.IsInvalid)
+                {
+                    myHandle.Dispose();
```

Comment:
What is the reason for disposing the handle when it is invalid?

Reply:
It would suppress the finalizer for the safe handle. It's on a path where the safe handle won't be used anymore, so it should ideally be disposed.